### PR TITLE
Enabled tuple merging in the executor

### PR DESCRIPTION
**Pull request by matan129**

### CODE REVIEW

File: storm-data-contracts/src/main/java/com/forter/contracts/BaseContractsBoltExecutor.java [hunk 1]
```diff
@@ -199,8 +202,20 @@ protected TInput transformInput(Object contract) throws JsonProcessingException
     protected void reportCacheStatus(Boolean status, Tuple tuple) {
     }
 
-    private void emit(Object id, Object contract, BasicOutputCollector collector) {
-        ArrayList<Object> tuple = Lists.newArrayList(id, transformOutput(contract));
+    /**
+     * Override this method for different merge strategies
+     */
+    protected List<Object> mergeTuples(List<Object> update, Tuple originalInput) {
+        return update;
```
Comment (itaifrenkel) [inline storm-data-contracts/src/main/java/com/forter/contracts/BaseContractsBoltExecutor.java]: please implement given assumptions

File: storm-data-contracts/src/main/java/com/forter/contracts/EnrichmentBolt.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Marks enrichment bolts that just alter and add to the tuple but do not "absorb" anything
+ */
+@Retention(RetentionPolicy.RUNTIME)
```
Comment (itaifrenkel) [inline storm-data-contracts/src/main/java/com/forter/contracts/EnrichmentBolt.java]: please add example in readme, one liner expaination

Comment (matan129) [inline storm-data-contracts/src/main/java/com/forter/contracts/EnrichmentBolt.java]: done

File: storm-data-contracts/src/main/java/com/forter/contracts/BaseContractsBoltExecutor.java [hunk 3]
```diff
@@ -199,8 +199,32 @@ protected TInput transformInput(Object contract) throws JsonProcessingException
...
+
+        for(Map.Entry<String, Object> entry : originalAttributes.entrySet()) {
+            finalAttributes.put(entry.getKey(), entry.getValue());
+        }
+
+        for(Map.Entry<String, Object> entry : newAttributes.entrySet()) {
+            finalAttributes.put(entry.getKey(), entry.getValue());
+        }
+
+        return Lists.newArrayList(update.get(0), finalAttributes);
```
Comment (itaifrenkel) [inline storm-data-contracts/src/main/java/com/forter/contracts/BaseContractsBoltExecutor.java]: update.put(1,finalAttributes)

File: storm-data-contracts/src/main/java/com/forter/contracts/BaseContractsBoltExecutor.java [hunk 4]
```diff
@@ -199,8 +199,32 @@ protected TInput transformInput(Object contract) throws JsonProcessingException
...
+    }
+
+    private void emit(Object id, Object contract, Tuple originalInput, BasicOutputCollector collector) {
+        List<Object> tuple = Lists.newArrayList(id, transformOutput(contract));
```
Comment (itaifrenkel) [inline storm-data-contracts/src/main/java/com/forter/contracts/BaseContractsBoltExecutor.java]: protected createTuple

File: storm-data-contracts/src/main/java/com/forter/contracts/BaseContractsBoltExecutor.java [hunk 5]
```diff
@@ -199,8 +199,32 @@ protected TInput transformInput(Object contract) throws JsonProcessingException
     protected void reportCacheStatus(Boolean status, Tuple tuple) {
     }
 
-    private void emit(Object id, Object contract, BasicOutputCollector collector) {
-        ArrayList<Object> tuple = Lists.newArrayList(id, transformOutput(contract));
+    /**
+     * Override this method for different merge strategies
+     */
+    protected List<Object> mergeTuples(List<Object> update, Tuple originalInput) {
```
Comment (itaifrenkel) [inline storm-data-contracts/src/main/java/com/forter/contracts/BaseContractsBoltExecutor.java]: enrichAttributes

File: storm-data-contracts/src/main/java/com/forter/contracts/BaseContractsBoltExecutor.java [hunk 6]
```diff
@@ -199,8 +199,32 @@ protected TInput transformInput(Object contract) throws JsonProcessingException
...
+        Map<String, Object> originalAttributes = (Map<String, Object>)originalInput.getValue(1);
+        Map<String, Object> newAttributes = (Map<String, Object>)update.get(1);
+        Map<String, Object> finalAttributes = new HashMap<>();
+
+        for(Map.Entry<String, Object> entry : originalAttributes.entrySet()) {
+            finalAttributes.put(entry.getKey(), entry.getValue());
```
Comment (itaifrenkel) [inline storm-data-contracts/src/main/java/com/forter/contracts/BaseContractsBoltExecutor.java]: putAll

**Pull request merged**